### DISCUSSION TITLE
[chpater14] 리팩터링

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/is-my-code-that-strange#121] [Chapter 13-14 (모델링, 리팩터링)] [answered]
p312. 테스트가 없는 코드에 테스트를 추가해서 안전하게 리팩터링할 수 있는 여러 자기 테크닉을 소개하며,
책에서는 2가지 방법은 소개하고 있습니다.

실제 업무에서 테스트가 없는 코드에 테스트 코드를 추가하여 리팩터링한 경험이 있을까요?
책에서 소개한 내용 말고 추가적으로 레거시 코드를 리팩터링할 때 경험했던 내용이 있다면 알려주세요~!

**Comment by firewoody237** [comment]
내부 비즈니스 로직이 많이 없는 API들을 마이그레이션한 적이 있는데요,
비즈니스 로직이 없기 때문에 'Service Layer - Repository Layer' 부분에 대해서만 JdbcTemplate을 사용하여
Insert -> 수행 -> Select -> Delete 쿼리로 테스트를 짜서 이관한 경험이 있습니다.

**Comment by corock** [comment, marked as answer]
최근에 주문서에서 사용할 수 있는 쿠폰과 적용할 수 없는 쿠폰 목록을 조회하는 API 를 개선한 적이 있습니다.
여건상 코드 전체 대신 부분적으로 리팩터링을 진행했고, 특히 저에게는 14.5 내용이 많이 와닿았습니다.
(기능 추가와 리팩터링 동시에 하지 않기, 작은 단계로 실시하기, 불필요한 사양은 제거 고려하기, ...)

아래는 리팩터링한 부분을 살짝 각색한 코드 스니펫입니다.

```java
public class OrderSheetCouponService {

    // 호출 메서드 리팩터링 진행
    private final AService aService;

    // 수정 시 영향도가 커서 그대로 둠
    private final BService bService;
    private final CService cService;

    // 신규 추가 (OCP)
    private final DComponent dComponent;
    private final EComponent eComponent;
    private final FComponent fComponent;

    public OrderSheetCouponResponse getOrderSheetCoupons(Long orderId) {
        // (...)

        Foo foo = aService.getFoo(..);      // 메서드 정의 영역 리팩터링
        dComponent.bar(..);                 // 신규 추가

        // (...)

        Baz baz = eComponent.getBaz(..);    // 신규 추가

        // (...)

        Qux qux = fComponent.qux(..);       // 신규 추가

        // (...)

        for (XxxCoupon xxxCoupon : xxxCoupons) {
            // (...)

            if (dComponent.Www(..)) {
                cService.makeXxx(..);       // AS-IS 유지
                cService.setYyy(..);        // AS-IS 유지

                // (...)

                fComponent.zzz(..);         // 신규 추가
            }
        }
    }

}
```

책에 소개된 것처럼 테스트가 없는 코드에 단위 테스트를 작성할 엄두는 안 나서 수정 범위가 적은 `aService` 의 `getFoo` 메서드나
새로 추가한 `dComponent`, `eComponent`, `fComponent` 에 대해서는 테스트를 추가하고 개선 작업을 진행했습니다.
다만, 새로 추가한 컴포넌트는 AS-IS 코드와 비교하는 것이 아니라 화면과의 비교나 정책 확인 등을 통해서 진행하고 QA 를 요청했네요.

그리고 직접 경험해본 적은 없지만 A/B testing 을 통해서도 API 를 리팩터링할 수 있는 것으로 알고 있습니다.
주로 UX 변경 내용을 확인할 때 자주 사용되는 테스팅 기법이지만 API 간 성능 비교를 할 때 적절하게 활용해 볼 수 있을 것 같습니다.
